### PR TITLE
ARROW-3972: [C++] Migrate to LLVM 7. Add option to disable using ld.gold

### PR DESCRIPTION
**Pull request by wesm**
All in all this wasn't too painful, and Gandiva seems to run fine (I wouldn't have expected otherwise).

I tested this locally on Ubuntu 18.10, and found that the LLVM libraries from apt fail to link with ld.gold (binutils 1.16), so I added an option to toggle ld.gold ON and OFF (it's now OFF by default; using `ld` by default should yield strictly fewer bugs / build failures). Not sure why that is

I will need some help testing and debugging the Crossbow and packaging tasks to make sure this doesn't break anything else before we merge it

### CODE REVIEW

**Comment by wesm** [comment]
@kszucs can you help with running / checking the crossbow tasks?

@xhochy what's involved in upgrading manylinux1 beyond changing the build_llvm.sh script?

**Comment by wesm** [comment]
@pravindra I wasn't sure whether this would be controversial but we need to move to LLVM 7 to be able to package Gandiva with the Python wheels in the next major release (0.13.0)

**Comment by xhochy** [comment]
> @xhochy what's involved in upgrading manylinux1 beyond changing the build_llvm.sh script?

Updating the `scripts/build_clang.sh` script. Otherwise that should be it.

**Comment by xhochy** [comment]
@wesm `ld.gold` problems are normally good indicators of real linking problems. What did you face?

**Comment by wesm** [comment]
I'll try building again later and post the errors here

**Comment by pravindra** [comment]
I'm pleasantly surprised that there are no gandiva issues :-).

Once the CI build stabilises, I will try your change with the crossbow task used by dremio. Can you please hold off on merging this change till then ?

**Comment by wesm** [comment]
Yes, of course

File: ci/travis_install_toolchain.sh [hunk 1]
```diff
@@ -25,7 +25,7 @@ if [ ! -e $CPP_TOOLCHAIN ]; then
     CONDA_PACKAGES=""
 
     if [ $ARROW_TRAVIS_GANDIVA == "1" ] && [ $TRAVIS_OS_NAME == "osx" ]; then
-        CONDA_PACKAGES="$CONDA_PACKAGES llvmdev=6.0.1"
+        CONDA_PACKAGES="$CONDA_PACKAGES llvmdev=$CONDA_LLVM_VERSION"
```
Comment (pitrou) [inline ci/travis_install_toolchain.sh]: Can use the syntax `llvmdev="7.0.*"` instead of hardcoding the bugfix level.

Comment (wesm) [reply ci/travis_install_toolchain.sh]: done

**Comment by wesm** [comment]
Here's what the ld.gold errors look like:

https://gist.github.com/wesm/c11c4ebd90be74f9e536f86d11f540c5

This was on Ubuntu Cosmic / 18.10

I'm going to try this on Ubuntu 14.04 also to see if the issue is somehow related to LLVM or whether it is particular to Ubuntu 18.10

**Comment by wesm** [comment]
The build worked on Ubuntu 14.04 using both ld and gold. Not sure where to go for here, this is not my expertise

I think the prudent thing is to disable ld.gold by default -- if there is a build configuration where it works well and reduces build times, then we can definitely turn it on

**Comment by xhochy** [comment]
> @xhochy do you want to take care of the LLVM 7 upgrade in this patch or in a different one?
> 
> Anything else you all want to take care of in this patch?

You mean in the manylinux1 container? Or in total? I would only do the former.

**Comment by pravindra** [comment]
@wesm I tried the crossbow builds used by dremio with your patch. I found two issues 

1. conda requires full version (llvmdev-7.0), whereas apt-get requires only major number (clang-7)
2. on mac, the script couldn't find clang-7. 
    - I couldn't find clang-7 on conda. so, for now, i just modified the script to switch to default clang if clang-7 is not found.

My changes are here : https://github.com/pravindra/arrow/commit/3da660bc8a01affcdf8a021802af04227030fb87

**Comment by xhochy** [comment]
@pravindra The `clang` package called `clangdev` on conda.

**Comment by wesm** [comment]
thanks @pravindra! I rebased and cherry picked your changes. I'll merge this once the build is once the build is passing

**Comment by pravindra** [comment]
@wesm, I can work on fixing the xenial build.

**Comment by wesm** [comment]
Linking to all the static libraries didn't make that error go away...

**Comment by wesm** [comment]
Thanks

**Comment by pravindra** [comment]
@wesm - the xenial error was happening because it was picking up clang/llvm from travis (pre-installed) rather than the one from apt-get. Using the one from apt-get fixed the issue. I also reverted the change to link only the required llvm libs.

https://github.com/wesm/arrow/pull/7

**Comment by wesm** [comment]
That's great, thank you.

**Comment by wesm** [comment]
Travis CI looking OK now. I just rebased to make sure we get a clean build. I'll then merge this and notify the mailing list about upgrading clang toolchains

This is good also because Visual Studio 2017 requires at least clang 7 -- I hit that when doing Windows testing

**Comment by wesm** [comment]
Appveyor build https://ci.appveyor.com/project/wesm/arrow/builds/22177695

**Pull request closed without merging**